subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/DefaultLocalComponentRegistry.java

```diff
@@ -18,74 +18,33 @@
-import org.gradle.internal.model.CalculatedValueContainer;
-import org.gradle.internal.model.CalculatedValueContainerFactory;
-import org.gradle.internal.model.ValueCalculator;
 
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
-
-public class DefaultLocalComponentRegistry implements LocalComponentRegistry, HoldsProjectState {
+/**
+ * Default implementation of {@link LocalComponentRegistry}.
+ */
+public class DefaultLocalComponentRegistry implements LocalComponentRegistry {
     private final BuildIdentifier thisBuild;
-    private final ProjectStateRegistry projectStateRegistry;
-    private final CalculatedValueContainerFactory calculatedValueContainerFactory;
-    private final LocalComponentProvider provider;
-    private final LocalComponentInAnotherBuildProvider otherBuildProvider;
-    private final Map<ProjectComponentIdentifier, CalculatedValueContainer<LocalComponentGraphResolveState, ?>> projects = new ConcurrentHashMap<>();
+    private final BuildTreeLocalComponentProvider componentProvider;
 
     public DefaultLocalComponentRegistry(
         BuildIdentifier thisBuild,
-        ProjectStateRegistry projectStateRegistry,
-        CalculatedValueContainerFactory calculatedValueContainerFactory,
-        LocalComponentProvider provider,
-        LocalComponentInAnotherBuildProvider otherBuildProvider
+        BuildTreeLocalComponentProvider componentProvider
     ) {
         this.thisBuild = thisBuild;
-        this.projectStateRegistry = projectStateRegistry;
-        this.calculatedValueContainerFactory = calculatedValueContainerFactory;
-        this.provider = provider;
-        this.otherBuildProvider = otherBuildProvider;
+        this.componentProvider = componentProvider;
     }
 
     @Override
     public LocalComponentGraphResolveState getComponent(ProjectComponentIdentifier projectIdentifier) {
-        ProjectState projectState = projectStateRegistry.stateFor(projectIdentifier);
         if (isLocalProject(projectIdentifier)) {
-            CalculatedValueContainer<LocalComponentGraphResolveState, ?> valueContainer = projects.computeIfAbsent(projectIdentifier, projectComponentIdentifier ->
-                calculatedValueContainerFactory.create(Describables.of("metadata of", projectIdentifier), new MetadataSupplier(projectState)));
-            // Calculate the value after adding the entry to the map, so that the value container can take care of thread synchronization
-            valueContainer.finalizeIfNotAlready();
-            return valueContainer.get();
+            return componentProvider.getLocalComponent(projectIdentifier);
         } else {
-            return otherBuildProvider.getComponent(projectState);
+            return componentProvider.getForeignComponent(projectIdentifier);
         }
     }
 
     private boolean isLocalProject(ProjectComponentIdentifier projectIdentifier) {
         return projectIdentifier.getBuild().equals(thisBuild);
```

Comment:
In practice this works perfectly fine since both of these identifiers do take the underling build path into account. Though for the sake of being explicit I can update this